### ISSUE TITLE
Alpha 3 release (2015-12-18)

### ISSUE THREAD

**Issue by dirk-thomas** [ros2/ros2#144]
**Feature development:**

* [x] Refactor rclcpp into a library @wjwwood 
  * [x] Separate testing of Connext static and dynamic: ros2/ros2#147 + review + squash (3 days)
  * [x] Update ros1_bridge code to leverage that rclcpp is a library @dirk-thomas ros2/ros1_bridge#20
    * @dirk-thomas ~~reduce memory requirement by splitting generated factories into separate source files~~ no further work necessary, the generated cpp file compiles and links in a reasonable time frame and without a significant memory usage

* [x] Design rcl interface (1 week) @wjwwood 
  * [ ] Refactor parts of rclcpp into rcl (2 week) (does not include refactoring rclcpp to use it)
    * [x] Focus on node, pub/sub, waitset, guard conditions @wjwwood 
    * [x] Include C message structures (1 week)
      * [x] get the generated C code to compile @dirk-thomas ros2/rosidl#49
      * [x] Add support for missing types (string, bounded arrays), provide alloc/free and init/fini functions (1 week) @dirk-thomas ros2/rosidl#52
      * [x] generate code for service request / response @dirk-thomas ros2/rosidl#54
      * [x] refactor type support to make rclcpp depend on a RMW implementation @dirk-thomas ros2/rmw_implementation#9
      * [x] Implement C typesupport @dirk-thomas ros2/rosidl#55
      * [ ] Add support for C data structures in all rmw implementations (2 weeks) @dirk-thomas 

* [ ] Real-time-safe intra-process messaging @jacquelinekay 
  * [ ] Thread scheduling and mutexes (2 week design phase + impl)
  * [ ] Test and implement allocator (1 week) 
    * [ ] ros2/realtime_support#16
  * [ ] Update demo (1 week)

* [x] Migrate TF support @tfoote https://github.com/ros2/ros2/issues/153
  * [x] Replace internal types (make truely ROS agnostic) (1/2 week design phase, +1/2 week implementation)
  * [x] Port ros2 wrapper programs/libraries (1 week) 

**Infrastructure:**

* [x] Cover all examples with unit tests: ros2/ros2#143 @esteve 
* [ ] Cover all demos with unit tests: https://github.com/ros2/demos/issues/43
* [x] Jenkins improvements:
  * [x] Use ccache ros2/ros2#145 @gerkey 
  * [x] Build binaries nightly ros2/ros2#146 @gerkey 
* [ ] Windows development guide ros2/ros2#148 @jacquelinekay
* [x] Windows default launching: https://github.com/ros2/ros2/issues/150
* [x] Windows multi launch job: https://github.com/ros2/ros2/issues/151

**Backlog:**

* *add tickets from the backlog when being handled*

* [x] cpplint ros2/ros2#12

**Background tasks:**

* [x] Support for FastRTPS @dirk-thomas (see ros2/ros2#124)

**Deferred until Alpha 4:**

* [ ] Component life-cycle in C++ as defined in the design article (with ROS-agnostic API) (2 weeks)
  * [ ] Introspection and orchestration APIs as ROS topics / services (2 weeks)
    * [ ] Use the orchestration API in launch to make tests deterministic (1 week)
  * [ ] Use C++ program to demonstrate life-cycle and orchestration in a tutorial/demo (1 week)
  * [ ] Use class_loader / pluginlib (1.5 weeks)
* [ ] Support services in C @wjwwood
* [ ] Python client library (estimate following rcl design meeting) @esteve 
* [ ] FreeRTPS improvements: RMW support @codebot 
  * [ ] Autogenerate deserialization (1 weeks)
  * [ ] Tie back into rmw (1 weeks)
  * [ ] Testing/CI for linux case (1 weeks)

**Comment by dirk-thomas** [comment]
We have deferred "Component life-cycle" until Alpha 4.

**Comment by dirk-thomas** [comment]
I have striked through the item "reduce memory requirement by splitting generated factories into separate source files". It looks like the current code the generated cpp file compiles and links in a reasonable time frame and without a significant memory usage. Therefore I don't think any further work is necessary right now.

Since I updated the content in the description this message is going to self destruct in 3, 2, 1 :bomb:
